### ISSUE TITLE
Rails 4.2.10 "exists?" always on master

### ISSUE THREAD

**Issue by camol** [instacart/makara#195]
Hi,

I just noticed a strange behaviour in makara. I would assume that all select queries go to slave. Especially such simply query which is being generated by `exists?` method in Rails. I would like to have more control over such simple queries execution. Here are the examples:

```
>> User.exists? #=> true
D, [2018-02-21T14:31:45.918264 #32775] DEBUG -- :   [master/1] User Exists (2.2ms)  SELECT  1 AS one FROM "users" LIMIT 1
```

```
>> User.connection.execute("SELECT  1 AS one FROM users LIMIT 1") #=> #<PG::Result:0x00007fd054d9c0c0 status=PGRES_TUPLES_OK ntuples=1 nfields=1 cmd_tuples=1>
D, [2018-02-21T14:32:31.968956 #32775] DEBUG -- :   [slave/1] (1.9ms)  SELECT  1 AS one FROM users LIMIT 1
```

Is this a bug or this is intended? At first glance it looks like a bug since the produced SQL queries are the same for both examples.

**Comment by bleonard** [comment]
Upon first glance, I don't know any reason why that would be the case. Want to make a test that fails an/or look into it?
There is a way to to override the regex that is used if you want to work around it.

**Comment by camol** [comment]
I will try to prepare a test case. I will probably need to prepare a dockerized Rails 4 app with postgresql to make sure all stays the same.

I know that there is a regexp you mentioned:
https://github.com/taskrabbit/makara/blob/e95d4303d350c9ab9dfc41b1e5c2743a1849b9c4/lib/active_record/connection_adapters/makara_abstract_adapter.rb#L112

but for the first example code the SQL that I get do not represent the query I think it should (some Rails internal querying) and the second actually is there. The funny thing is that even when the `sql` variable is what I expect for second example [here](https://github.com/taskrabbit/makara/blob/e95d4303d350c9ab9dfc41b1e5c2743a1849b9c4/lib/active_record/connection_adapters/makara_abstract_adapter.rb#L184) I still get `false` but it lands on slave 😄.

**Comment by bleonard** [comment]
I tried this and it hits the slave: https://github.com/taskrabbit/makara/pull/197
I also tried in our app.
```
Loading development environment (Rails 4.2.10)
irb(main):001:0> User.exists?
Dalli::Server#connect 127.0.0.1:11211
  [slave/1] User Exists (2.8ms)  SELECT  1 AS one FROM `users` LIMIT 1
=> true
```

Not sure of the differences at play.

**Comment by camol** [comment]
@bleonard it looks like this is somehow connected with the version of the Rails, on Rails 3 I did not have this issue. [Here](https://github.com/camol/makara_test_app) is the test app which I prepared using docker-compose. You can try to clone the repo and do:
1. `docker-compose build`
2. `docker-compose run --rm app bash`
3. `bundle exec rake db:create db:migrate`
4. `bundle exec rails c`
5. `User.exists?`
-> [master/1] User Exists (0.9ms)  SELECT  1 AS one FROM "users" LIMIT 1
6. `ActiveRecord::Base.connection.execute("SELECT  1 AS one FROM users LIMIT 1")`
-> [slave/1] (1.9ms)  SELECT  1 AS one FROM users LIMIT 1

EDIT:
I did some debugging on both Rails 3 and Rails 4 with makara by adding some debugging messages here:

```
    def _appropriate_pool(method_name, args)
      # the args provided absolutely need master
      if needs_master?(method_name, args)
        p "HERE 1"
        stick_to_master(method_name, args)
        @master_pool

      # in this context, we've already stuck to master
      elsif Makara::Context.get_current == @master_context
        p "HERE 2"
        @master_pool

      elsif previously_stuck_to_master?
        p "HERE 3"

        # we're only on master because of the previous context so
        # behave like we're sticking to master but store the current context
        stick_to_master(method_name, args, false)
        @master_pool

      # all slaves are down (or empty)
      elsif @slave_pool.completely_blacklisted?
        p "HERE 4"
        stick_to_master(method_name, args)
        @master_pool

      elsif in_transaction?
        p "HERE 5"
        @master_pool

      # yay! use a slave
      else
        p "HERE 6"
        @slave_pool
      end
    end
```

I see that for both envs and both queries end up in the "HERE 6" so they in fact fallback to slave_pool but somehow on Rails 4 this becomes the master by looking at the final log.

I also jumped into this method:

```
      def needs_master?(method_name, args)
        sql = coerce_query_to_sql_string(args.first)
        p method_name
        p args
        p sql
        return true if sql_master_matchers.any?{|m| sql =~ m }
        return false if sql_slave_matchers.any?{|m| sql =~ m }
        true
      end
```

in makara abstract adapter and for Rails 4 I see this:

```
>> User.exists? #=> true
:exec_query
["              SELECT a.attname, format_type(a.atttypid, a.atttypmod),\n                     pg_get_expr(d.adbin, d.adrelid), a.attnotnull, a.atttypid, a.atttypmod\n                FROM pg_attribute a LEFT JOIN pg_attrdef d\n                  ON a.attrelid = d.adrelid AND a.attnum = d.adnum\n               WHERE a.attrelid = '\"users\"'::regclass\n                 AND a.attnum > 0 AND NOT a.attisdropped\n               ORDER BY a.attnum\n", "SCHEMA"]
"              SELECT a.attname, format_type(a.atttypid, a.atttypmod),\n                     pg_get_expr(d.adbin, d.adrelid), a.attnotnull, a.atttypid, a.atttypmod\n                FROM pg_attribute a LEFT JOIN pg_attrdef d\n                  ON a.attrelid = d.adrelid AND a.attnum = d.adnum\n               WHERE a.attrelid = '\"users\"'::regclass\n                 AND a.attnum > 0 AND NOT a.attisdropped\n               ORDER BY a.attnum\n"
"HERE 6"
D, [2018-02-27T15:13:45.381256 #50690] DEBUG -- :   [master/1] User Exists (2.0ms)  SELECT  1 AS one FROM "users" LIMIT 1
```

for Rails 3 it gives me this:

```
>> User.exists? #=> true
:exec_query
["SELECT  1 AS one FROM \"users\"  LIMIT 1", "User Exists", []]
"SELECT  1 AS one FROM \"users\"  LIMIT 1"
"HERE 6"
  [slave/1] User Exists (3.8ms)  SELECT 1 AS one FROM "users" LIMIT 1
```

**Comment by johnwu96822** [comment]
We are having the same issue as well. The stack trace showed that `exists?` call ended up in proxy.rb `method_missing` and calls `any_connection`. Then any_connection will try master first. We are on Rails 4.2.8.

**Comment by camol** [comment]
@bleonard did you manage to use my example app and replicate the issue?

**Comment by bleonard** [comment]
I found that the postgres gem overrides some methods and we need to hijack them.
PR here: https://github.com/taskrabbit/makara/pull/199

**Comment by camol** [comment]
@bleonard when do you think it will land on master?

**Comment by bleonard** [comment]
Testing it out now on master.
Try this if you like: `gem 'makara', git: 'https://github.com/taskrabbit/makara.git', tag:'v0.3.10'`

**Comment by johnwu96822** [comment]
@bleonard FYI we have this issue witth MySQL, so I hope the fix will work for MySQL as well.

**Comment by bleonard** [comment]
@johnwu96822 can you see if this tag fixes it? I wouldn't have expected to see the issue on mysql given what I learned during this process.

**Comment by camol** [comment]
@bleonard thx for the fix, it works for me.

**Comment by johnwu96822** [comment]
Hi @bleonard Locally it seems fine, even for 0.3.9. Unfortunately I won't be able to test it any time soon in production.

For 0.3.9 in production, this is what we see in the stack trace:
```
/gems/makara-0.3.9/lib/makara/proxy.rb:104 in method_missing
/gems/activerecord-4.2.8/lib/active_record/relation/finder_methods.rb:302 in exists?
```
Basically the `exists?`  calls the `select_value` on the connection https://github.com/rails/rails/blob/4-2-stable/activerecord/lib/active_record/relation/finder_methods.rb#L302

And then perhaps `select_value` triggered method_missing and calls any_connection, which always tries master first https://github.com/taskrabbit/makara/blob/v0.3.9/lib/makara/proxy.rb#L104

**Comment by bleonard** [comment]
Released as part of 0.3.10: https://rubygems.org/gems/makara/versions/0.3.10
